crates/matrix-sdk/src/client/mod.rs

```diff
@@ -253,10 +253,26 @@ impl Client {
     #[cfg(feature = "appservice")]
     pub async fn receive_transaction(
         &self,
-        _transaction_id: &TransactionId,
+        transaction_id: &TransactionId,
         sync_response: sync_events::v3::Response,
     ) -> Result<()> {
-        // TODO: transaction id checking, see PR #560
+        const TXN_ID_KEY: &[u8] = b"appservice.txn_id";
+
+        let store = self.store();
+        let store_tokens = store.get_custom_value(TXN_ID_KEY).await?;
+        let mut txn_id_bytes = transaction_id.as_bytes().to_vec();
+        if let Some(mut store_tokens) = store_tokens {
+            // The data is separated by a NULL byte.
+            let mut store_tokens_split = store_tokens.split(|x| *x == b'\0');
+            if store_tokens_split.any(|x| x == transaction_id.as_bytes()) {
+                return Ok(());
```

Comment:
A short comment explaining why we return early here would be good. E.g.

```
We already encountered this transaction id before, so we exit early instead of processing further.
```